src/query/expression/src/row/row_converter.rs

```diff
@@ -0,0 +1,238 @@
+                            })
+                    } else {
+                        col.as_variant()
+                            .unwrap()
+                            .iter()
+                            .zip(lengths.iter_mut())
+                            .for_each(|(bytes, length)| {
+                                *length += variable::encoded_len(bytes, false) as u64
+                            })
+                    }
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        let mut offsets = Vec::with_capacity(num_rows + 1);
+        offsets.push(0);
+
+        // Comments from apache/arrow-rs:
+        // We initialize the offsets shifted down by one row index.
+        //
+        // As the rows are appended to the offsets will be incremented to match
+        //
+        // For example, consider the case of 3 rows of length 3, 4, and 6 respectively.
+        // The offsets would be initialized to `0, 0, 3, 7`
+        //
+        // Writing the first row entirely would yield `0, 3, 3, 7`
+        // The second, `0, 3, 7, 7`
+        // The third, `0, 3, 7, 13`
+        //
+        // This would be the final offsets for reading
+        //
+        // In this way offsets tracks the position during writing whilst eventually serving
+        // as identifying the offsets of the written rows
+        let mut cur_offset = 0_u64;
+        for l in lengths {
+            offsets.push(cur_offset);
+            cur_offset = cur_offset.checked_add(l).expect("overflow");
+        }
+
+        let buffer = vec![0_u8; cur_offset as usize];
+
+        StringColumnBuilder::from_data(buffer, offsets)
+    }
+}
+
+#[inline(always)]
+pub(super) fn null_sentinel(nulls_first: bool) -> u8 {
+    if nulls_first { 0 } else { 0xFF }
+}
+
+fn encode_column(out: &mut StringColumnBuilder, column: &Column, asc: bool, nulls_first: bool) {
+    let validity = column.validity();
+    let column = column.remove_nullable();
+    match column {
+        Column::Null { .. } => {}
+        Column::Boolean(col) => fixed::encode(out, col, validity, asc, nulls_first),
+        Column::Number(col) => {
+            with_number_type!(|NUM_TYPE| match col {
+                NumberColumn::NUM_TYPE(c) => {
+                    fixed::encode(out, c, validity, asc, nulls_first)
+                }
+            })
+        }
+        Column::Decimal(col) => {
+            with_decimal_type!(|DECIMAL| match col {
+                DecimalColumn::DECIMAL(c, _) => {
+                    fixed::encode(out, c, validity, asc, nulls_first)
+                }
+            })
+        }
+        Column::Timestamp(col) => fixed::encode(out, col, validity, asc, nulls_first),
+        Column::Date(col) => fixed::encode(out, col, validity, asc, nulls_first),
+        Column::String(col) => variable::encode(out, col.iter(), validity, asc, nulls_first),
+        Column::Variant(col) => variable::encode(out, col.iter(), validity, asc, nulls_first),
```

Comment:
treating `Variant` as `String` may lead to unexpected results, we can treat the `Variant` as an unsupported type and sort it in other methods.

Reply:
Ok, we can only support one variant order-by column currently.